src/core/simulation.jl

```diff
@@ -884,97 +923,142 @@ function update_stage!(
-            _execute!(sim; kwargs...)
+        open_func(get_store_dir(sim), "w") do store
+            # TODO: return file name for hash calculation instead of hard code
+            Logging.with_logger(logger) do
+                TimerOutputs.reset_timer!(RUN_SIMULATION_TIMER)
+                TimerOutputs.@timeit RUN_SIMULATION_TIMER "Execute Simulation" begin
+                    _execute!(sim, store; kwargs...)
+                end
+                @info ("\n$(RUN_SIMULATION_TIMER)\n")
+                set_simulation_status!(sim, RunStatus.SUCCESSFUL)
+                log_cache_hit_percentages(store)
+            end
         end
+    catch e
+        # TODO: Add Fallback when run_stage fails
+        set_simulation_status!(sim, RunStatus.FAILED)
+        @error "simulation failed" exception = (e, catch_backtrace())
     finally
         unregister_recorders!(sim.internal)
         close(logger)
     end
+    compute_file_hash(get_store_dir(sim), HDF_FILENAME)
+    return get_simulation_status(sim)
 end
 
-function _execute!(sim::Simulation; kwargs...)
-    if sim.internal.reset
-        sim.internal.reset = false
-    elseif sim.internal.reset == false
-        error("Re-build the simulation")
+function _execute!(
+    sim::Simulation,
+    store;
+    cache_size_mib = 1024,
+    exports = nothing,
+    kwargs...,
+)
+    @assert !isnothing(sim.internal)
+    set_simulation_status!(sim, RunStatus.RUNNING)
+    execution_order = get_execution_order(sim)
+    steps = get_steps(sim)
+    num_executions = steps * length(execution_order)
+    store_params = _initialize_stage_storage!(sim, store, cache_size_mib)
+    initialize_optimizer_stats_storage!(store, num_executions)
+    status = RunStatus.RUNNING
+    if exports !== nothing
+        if !(exports isa SimulationResultsExport)
+            exports = SimulationResultsExport(exports, store_params)
+        end
+
+        if exports.path === nothing
+            exports.path = get_results_dir(sim)
+        end
     end
-    system_to_file = get(kwargs, :system_to_file, true)
-    sim.internal === nothing &&
-        error("Simulation not built, build the simulation to execute")
-    TimerOutputs.reset_timer!(RUN_SIMULATION_TIMER)
-    TimerOutputs.@timeit RUN_SIMULATION_TIMER "Execute Simulation" begin
-        execution_order = get_execution_order(sim)
-        for step in 1:get_steps(sim)
-            TimerOutputs.@timeit RUN_SIMULATION_TIMER "Execution Step $(step)" begin
-                println("Executing Step $(step)")
-                IS.@record :simulation_status SimulationStepEvent(
+
+    progress_bar = ProgressMeter.Progress(num_executions, 1)
+    for step in 1:steps
+        TimerOutputs.@timeit RUN_SIMULATION_TIMER "Execution Step $(step)" begin
+            IS.@record :simulation_status SimulationStepEvent(
+                get_current_time(sim),
+                step,
+                "start",
+            )
+            for (ix, stage_number) in enumerate(execution_order)
```

Comment:
Everything in this for loop should go in a separate function that takes the stage. We'll need this to enable the SIIP-MODSIM and Helics style extensions that won't be setting up a sim-store object.